design/replication-policy.md

```diff
@@ -0,0 +1,64 @@
+**FoundationDB Replication Policy System**
+
+FoundationDB uses a composable **replication policy system** to determine how data replicas are distributed across physical infrastructure. This system enables 
+the database to enforce constraints that ensure **fault tolerance**, **availability**, and **performance** while abstracting details from end users.
+
+Shards, representing key ranges, are replicated across some number of storage servers (the shard's "team"). FDB maintains a "shard map" data structure to map 
+shards to the team. More information on these terms can be found in [data-distributer-internals.md](http://data-distributer-internals.md)
+
+The commit proxy uses the shard map to derive which log servers to send committed data to. Each SS has a "buddy" with a single corresponding log server. The 
+buddy is found by taking the SS id modded by the number of tLogs in the dc.
+
+When a mutation is committed, the key(s) it modifies will logically reside within one or more shards, which physically are replicated over the SS in their team. Each such SS has a buddy 
+log server, and in the simple case, the CP forwards writes to those logs.
+
+For example, suppose there were 4 log servers and 10 storage servers on a cluster configured with 3 replicas, and a shard's team resides on SS 0, 3, 6. Given 
+there are 4 log servers total, the destination log servers would be {0%0, 3%4, 6%4}, or {0,3,2}.
+
+Storage servers: **0** 1 2 **3** 4 5 **6** 7 8 9
+
+TLogs: **0** 1 **2** **3**
+
+The "buddy" associations between the SS and LS are dynamic.
+
+- The number of SS in the cluster can change within an epoch.
+- Shards move between SS as data is distributed.
+- New shards can be created.
+
+This means the destination team must be recalculated on each commit.
+
+The team of log servers must meet the requirements of the replication policy. If it does not, additional logs are chosen. This is performed by "selectReplicas".
+
+The selectReplicas function accepts a list of available servers and already chosen "also" servers (if any). It will check if the already chosen servers meet the 
+replication policy. If they do not, the function will select additional servers at random out of the available servers in order to meet the policy. Because the 
+shard map has replicas for all key ranges,the set of "also" servers passed to selectReplica usually already satisfies the policy, but there are exceptions.
+
+- There are code paths where the shard map is not consulted, for example, when a tag is added to a transaction as a consequence of a metadata mutation.
+- Two SS may have the same buddy, in which case a random server must be chosen.
+
+Though users interact with only a few predefined redundancy modes (double, triple, etc.), FoundationDB internally supports a **rich set of compositional 
```

Comment:
Nice to see this description!

Reply:
In this case the "also" servers would not meet the policy. The code would have to search the set of "available" servers to find a combination that works. If more than one server from the available list meets the requirements, a random one from that set would be selected.

Reply:
This description of policies is super useful!